release-21.1: clusterversion: allow v21.1-124 to v21.1 downgrade

dt opened the pull request:
This softens the hard rule that cluster versions must be greater than prior cluster versions,
when opening a store, joining a cluster, or changing the setting, to add a special-case
exemption for 21.1-124, which was backported to release-21.1 and released in v21.1.8.

This means a node with this change, if it were released as 21.1.9, would be happy to allow
reducing the cluster setting fro 21.1-124 to 21.1. It also means a subsequent change could
change the binary version back to 21.1 and still be able to start on stores or join clusters
where the cluster version was previously bumped to 21.1-124.

Release note (bug fix): add backwards compatibility between 21.1.x cluster version and 21.1.8 cluster version.

dt committed:
clusterversion: allow v21.1-124 to v21.1 downgrade

This softens the hard rule that cluster versions must be greater than prior cluster versions,
when opening a store, joining a cluster, or changing the setting, to add a special-case
exemption for 21.1-124, which was backported to release-21.1 and released in v21.1.8.

This means a node with this change, if it were released as 21.1.9, would be happy to allow
reducing the cluster setting fro 21.1-124 to 21.1. It also means a subsequent change could
change the binary version back to 21.1 and still be able to start on stores or join clusters
where the cluster version was previously bumped to 21.1-124.

Release note (bug fix): add backwards compatibility between 21.1.x cluster version and 21.1.8 cluster version.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69894)
<!-- Reviewable:end -->

dt commented:
I think with something like this plus #69886, a final follow-up change could come in and delete `NewRetryableRangefeedErrors` and make `binaryVersion` go back to being `V21_1`?

ajwerner reviewed (commented):
I like it. Can you add a unit test that starts at `V21_1` and then sets the version up to `21Dot1Dot8` and then back down? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69894#-:-Mj1Fh4P1kr8vTA8Edm4:b-j2k64n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

dt commented:
Good call on test; it caught several cases of `Less()` checks I'd missed. Fixed those and included the test in updated diff.

knz commented on pkg/clusterversion/cockroach_versions.go:
```diff
@@ -571,3 +571,17 @@ func listBetweenInternal(from, to ClusterVersion, vs keyedVersions) []ClusterVer
+)
+
+// Is21Dot1Dot8Equiv checks if the passed versions are the two cluster versions
+// used by released 21.1 versions, that is, v21.1 an v21.1-124, which was added
```
And

knz reviewed (commented):
1. This also needs a roach test that validates that the downgrade actually works with old binaries.

2. You could sprinkle this change with "TODO" annotations that suggest the special case should be removed when version n+2 (22.2?) Is released.

dt committed:
roachtest: add revert to 21.1.0 to auto-upgrade test

Release note: none.

dt commented:
Added a roachtest that rolls back to 21.1 and then adds a 21.1.0 node.

As far as the TODOs, I don't think we need them since this is going to the 21.1 branch only.

knz reviewed (approved):
> Reviewed 4 of 5 files at r2, 4 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69894)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @irfansharif)



<!-- Sent from Reviewable.io -->

knz commented:
CI is unhappy about the new test though.

dt commented:
Ah, that's the unit test I had initially and _was_ green on GI before I added the additional roachtest, and was is passing when I run it locally, so must have uncovered an ability to flake, specifically if the test moves too quickly, after adding a node, to then adjust the version, before that node starts up? I guess TestCluster.Add doesn't wait until the new node is fully up, so I added the same liveness wait that we do elsewhere.

dt commented:
@irfansharif Can I get your eyes on the migration manager / RPC changes?

dt commented:
I'm a bit at a loss here as to why the migration manager is unhappy:
I'm doing `tc.AddAndStartServer` then `tc.WaitForNodeStatuses(t)` and `tc.WaitForNodeLiveness(t)` before a `SET CLUSTER SETTING version` but still:
```
n1: found liveness
n2: found liveness
n3: found liveness
manager_external_test.go:518:
  Error Trace:  manager_external_test.go:518
  Error:        Received unexpected error:
                pq: n3 required, but unavailable
  Test:         Test21Dot1Dot8
```
@irfansharif any ideas?

irfansharif reviewed (commented):
RPC changes LGTM.

dt commented:
@irfansharif do you have any idea why I needed the SucceedsSoon in the end of `Test21Dot1Dot8` even though I had the WaitForLiveness call above it? I kept getting CI flakes from the migration manager saying it needed n3 but n3 was unavailable (though I could never repro locally)

irfansharif commented:
Looks like tc.WaitForNodeLiveness simply waits for the liveness record to get created. When n3 attempts to join the cluster, and reaches out to n1 (or some other bootstrapped member), the record is created by n1: https://github.com/cockroachdb/cockroach/blob/d10b3a5badf25c9e19ca84037f2426b03196b2ac/pkg/server/node.go#L1367-L1377

There's probably some small delay between the record being created and n3 hearing about a successful Join RPC attempt and starting up for realsies. Maybe we want tc.WaitForNodeLiveness to instead wait for the first heartbeat (which would happen after n3 fully starts up).
